### ISSUE TITLE
Page Flags - Redaction App

### ISSUE THREAD

**Issue by KyEggleston** [bcgov/foi-flow#3413]
* As an IAO analyst
* I want to add page flags to pages 
* so that I can mark the status of each page in a records package

**Assumptions & Scope**	
What are the assumptions for this story?

Only one page flag can be selected for each page. 
All icons should be free from FontAwesome.

What is IN scope?
Page flags
Page flags tooltip

What is NOT in scope?
Page flag filtering
Total filtered pages counter
Percentage of filtered pages vs total pages counter
Filtered Pane views/functionality

**Acceptance Criteria**	

**Scenario 1: Page Flag Drop Down**
* GIVEN an IAO analyst is in the redaction app 
* WHEN they right-click a page in the left hand 'Organize By' pane
* THEN a page flag drop down will appear 
(Add design here)

**Scenario 2: Select Page Flag - Not Consult**
* GIVEN an IAO analyst has right-clicked a page in the left hand 'Organize By' pane 
* WHEN the analyst left-clicks a page flag from the drop down
* AND the page flag is not Consult
* THEN the drop down will close
* AND the page flag will be saved to the page

**Scenario 3: Select Page Flag - Consult**
* GIVEN an IAO analyst has right-clicked a page in the left hand 'Organize By' pane 
* WHEN the analyst hovers over the 'Consult' page flag in the drop down
* THEN a secondary drop down containing all ministry abbreviations (XXX) 
* AND an additional option at the bottom of the list titled 'Other Public Body', will be available
(Add design here)

**Scenario 4: Select Ministry - Consult**
* GIVEN an IAO analyst is hovered over 'Consult' in the page flag drop down
* AND the secondary drop down is displayed
* WHEN the analyst left-clicks a ministry abbreviation (XXX)
* THEN the drop downs will close, and the page flag will be saved to the page

**Scenario 5: Select Other Public Body - Consult**
* GIVEN an IAO analyst is hovered over 'Consult' in the page flag drop down
* AND the secondary drop down is displayed
* WHEN the analyst left-clicks 'Other Public Body'
* THEN the drop downs will close
* AND a modal will open with an open text field, and the options to 'save' or 'cancel'
* AND the 'save' option will be grey and inactive
* AND the 'cancel' option will be blue and active
(Add design here)

**Scenario 6: Other Public Body Modal - Enter Name**
* GIVEN an IAO analyst is on the 'Other Public Body' modal
* WHEN the analyst has entered any value (>=1 character) in the open text field
* THEN the 'save' and 'cancel' options will be blue and active

**Scenario 7: Other Public Body Modal - Save**
* GIVEN an IAO analyst is on the 'Other Public Body' modal
* AND they have entered any value (>=1 character) in the open text field
* WHEN the analyst left-clicks 'save'
* THEN the modal will close
* AND the page flag will be saved to the page

**Scenario 8: Other Public Body Modal - Cancel**
* GIVEN an IAO analyst is on the 'Other Public Body' modal
* AND they have entered any, or no, value in the open text field
* WHEN the analyst left-clicks 'cancel'
* THEN the modal will close
* AND the page flag will not be saved

**Scenario 9: Page Flag Iconography - Page Flag **
* GIVEN an IAO analyst has selected a page flag for a given page
* WHEN the IAO analyst sees the page name/number in the left hand 'Organize By' pane 
* THEN the page flag iconography will appear to the left of the page name
(Insert design here)

**Scenario 10: Page Flag Iconography - No Page Flag**
* GIVEN an IAO analyst has not selected a page flag for a given page
* WHEN the IAO analyst sees the page name/number in the left hand 'Organize By' pane 
* THEN no page flag iconography will appear next to the page name/number

**Scenario 11: Page Flag Iconography -  Non-Consult Tooltip **
* GIVEN an IAO analyst has selected a page flag that is not 'Consult' for a given page
* AND they see the page name/number in the left hand 'Organize By' pane with the corresponding page flag iconography
* WHEN the analyst hovers over the page flag iconography
* THEN a tooltip displaying the page flag name will be displayed 
(Add design here)

**Scenario 12: Page Flag Iconography - Consult Tooltip **
* GIVEN an IAO analyst has selected the 'Consult' page flag for a given page
* AND they see the page name/number in the left hand 'Organize By' pane with the corresponding page flag iconography
* WHEN the analyst hovers over the page flag iconography
* THEN a tooltip will be displayed with the text in the format of 'Consult - [open text field data]' 
(Add design here)

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by KyEggleston** [comment]
@lmullane @JHarrietha-AOT - please review, especially ACs 5-8, and 12. I'm not sure how we want handle the 'Other Public Body' consult option, where we can enter values into an open text field. It may be awkward if they have an open text field in the dropdown. May need to confirm how this will look, and I can then update the ACs after. 

@JHarrietha-AOT - feel free to add designs to ACs (I tried to indicate where they may be needed). 

cc: @abin-aot

**Comment by lmullane** [comment]
Nice work, @KyEggleston 

Made a few minor edits and one of significance - consult options should be Org Codes. That way the consult pages can be routed to the correct ministry queue for ministries that have more than one queue.

One challenge with scenario 12 is showing the Other public body (i.e. free text) may cause display issues if they enter a free text over a 5-6 characters. Options are to:

1.  Write another scenario that enforces a character limit in the free text field. Perhaps in future we have suggestive text based on previous entries. For example, if VAN was used for City of Vancouver, when the user starts writing VA...the system could suggest VAN.

2. When the user hovers over the consult icon only display 'Consult - Other' if the public body does not have an Org Code. And display Consult - Org Code for those that do.

What do you think, @JHarrietha-AOT, @KyEggleston, @m-prodan ?

**Comment by KyEggleston** [comment]
Added ACs for the completion percentage counter (calculated based on page flags)

**Comment by KyEggleston** [comment]
> Nice work, @KyEggleston
> 
> Made a few minor edits and one of significance - consult options should be Org Codes. That way the consult pages can be routed to the correct ministry queue for ministries that have more than one queue.
> 
> One challenge with scenario 12 is showing the Other public body (i.e. free text) may cause display issues if they enter a free text over a 5-6 characters. Options are to:
> 
> 1. Write another scenario that enforces a character limit in the free text field. Perhaps in future we have suggestive text based on previous entries. For example, if VAN was used for City of Vancouver, when the user starts writing VA...the system could suggest VAN.
> 2. When the user hovers over the consult icon only display 'Consult - Other' if the public body does not have an Org Code. And display Consult - Org Code for those that do.
> 
> What do you think, @JHarrietha-AOT, @KyEggleston, @m-prodan ?

I would lean towards to option 2 - I wasn't aware that IAO had org codes for outside of core government, so it was my assumption this would need open text.

**Comment by lmullane** [comment]
Option 2 makes sense to me too, @KyEggleston

**Comment by JHarrietha-AOT** [comment]
@KyEggleston Looks great. 

I made one change which is the dropdowns are all context menus not drop down (a drop down implys that there in a selection box and options drop down where a context menu appears overtop on something on a click) So i have updated that throughout.

As for the comment above I would agree with option 2, know the vast range of consults possible I don't think it is wise to enforce a character limit that might not make sense to others.

@KyEggleston @lmullane Thinking it could be helpful if when a users adds a new consult name that is automatically appears at the end of the consult list for that specific request. Just to avoid the user needing to type it multiple times and potentially spelling it differently and them not compiling to be export together easily.

**Comment by lmullane** [comment]
@JHarrietha-AOT, thanks for the review.

Let's discuss how to handle the new consult names at Release planning. Not sure we need it for MVP, but open to discussion.

**Comment by m-prodan** [comment]
Note update AC to include scenarios to cover shift clicks/control keys for multi-select (note: control click may be more effort, dev team needs to investigate)

**Comment by m-prodan** [comment]
@aparna-aot @sumathi-thirumani-aot - Jackie will be following up, but confirming we discussed and agreed on the following:

* When user creates an "add other" that consult location will now appear at the bottom of the list. Note - this should only persist at the file level, and not carry across to other requests. This will minimize user error and ensure consistency.
* Character limit for other public body should be 100 characters
* When public bodies have been selected, they will appear to the right of the page number indicating the three letter code for Internal consultations, and the first 10 characters for the "other"

cc @JHarrietha-AOT @lmullane

**Comment by m-prodan** [comment]
Updated AC to be consistent with design: should be Page Left Off, not Left Off Here

**Comment by lmullane** [comment]
As discussed at standup, added an "In Progress" page flag to be used by an analyst when they have started but not completed a page.

The auto-flag of 'In Progress' will come in a different story.

**Comment by m-prodan** [comment]
@aparna-aot - as discussed, let us put a pause on multi-select. Since it is not already in the AC, and due to other considerations, we should proceed as is. Once Abin's task is done for page counts and this can be QA'd - please proceed as is done, and disregard my comment about having multi-select.

**Comment by aparna-aot** [comment]
Noted. Thanks @m-prodan.

**Comment by aparna-aot** [comment]
Scenario 4 : Instead of Hover , implemented left click and confirmed with @JHarrietha-AOT as well.

**Comment by KyEggleston** [comment]
@aparna-aot 

Only thing I noticed is that for Scenario 11 and 12, no tooltips appear when hovering over the page flag iconography. 

![image](https://user-images.githubusercontent.com/117312323/226476705-3b70ee16-1f9a-4f34-99ec-7689a35912f3.png)

**Comment by aparna-aot** [comment]
@KyEggleston - I can update that soon & will let u know.

**Comment by aparna-aot** [comment]
@KyEggleston Tooltips added. You can verify now.Thanks!

**Comment by KyEggleston** [comment]
Thanks @aparna-aot! - All ACs pass, and I'll close ticket.
